include/unifex/any_sender_of.hpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/any_unique.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/get_stop_token.hpp>
+#include <unifex/inplace_stop_token.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+namespace _any {
+
+using _operation_state =
+    any_unique_t<overload<void(this_&) noexcept>(start)>;
+
+template <typename... Values>
+struct _rec_ref {
+  struct type;
+};
+
+template <typename... Values>
+struct _rec_ref<Values...>::type {
+  template <typename Op>
+  type(inplace_stop_token st, Op* op)
+    : op_(op)
+    , set_value_fn_([](void* op, Values&&... values) {
+        static_cast<Op*>(op)->stopTokenAdapter_.unsubscribe();
+        unifex::set_value(
+            std::move(static_cast<Op*>(op)->rec_),
+            (Values&&) values...);
+      })
+    , set_error_fn_([](void* op, std::exception_ptr e) noexcept {
+        static_cast<Op*>(op)->stopTokenAdapter_.unsubscribe();
+        unifex::set_error(
+            std::move(static_cast<Op*>(op)->rec_),
+            (std::exception_ptr&&) e);
+      })
+    , set_done_fn_([](void* op) noexcept {
+        static_cast<Op*>(op)->stopTokenAdapter_.unsubscribe();
+        unifex::set_done(
+            std::move(static_cast<Op*>(op)->rec_));
+      })
+    , st_(st)
+  {}
+
+  void set_value(Values... values) && {
+    set_value_fn_(op_, (Values&&) values...);
+  }
+  void set_error(std::exception_ptr e) && noexcept {
+    set_error_fn_(op_, (std::exception_ptr&&) e);
+  }
+  void set_done() && noexcept {
+    set_done_fn_(op_);
+  }
+
+private:
+  friend inplace_stop_token tag_invoke(tag_t<get_stop_token>, const type& self) {
+    return self.st_;
+  }
+
+  void *op_;
+  void (*set_value_fn_)(void*, Values&&...);
+  void (*set_error_fn_)(void*, std::exception_ptr) noexcept;
+  void (*set_done_fn_)(void*) noexcept;
```

Comment:
@lewissbaker @kirkshoop Would you prefer if I added an indirection and stored only a vtable ptr here? I think for 3 functions, this is better, but once we start adding type-erased receiver queries, a vtable is going to look attractive. We could even optimize for the common(?) case and leave the `set_value_fn_` pointer in the object but move `set_error_fn_` and `set_done_fn_` into a vtable.

Reply:
The any_unique implementation makes use of a vtable_holder_t type-alias that decides based on the number of vtable entries whether to store entries inline or externally.

Reply:
Additionally, I would not choose to pessimize set_done or set_error. There are cases where these are the hot path.